agent/consul/state/catalog.go

```diff
@@ -4483,11 +4532,24 @@ func (s *Store) ServiceTopology(
 		maxIdx = idx
 	}
 
-	var downstreamSources = make(map[string]string)
+	downstreamSources := make(map[string]string)
 	for _, dn := range downstreamNames {
 		downstreamSources[dn.String()] = structs.TopologySourceRegistration
 	}
 
+	idx, downstreamGWs, err := s.downstreamGatewaysForServiceTxn(tx, sn)
+	if err != nil {
+		return 0, nil, err
+	}
+	if idx > maxIdx {
+		maxIdx = idx
+	}
+
+	for _, dn := range downstreamGWs {
+		downstreamNames = append(downstreamNames, dn)
+		downstreamSources[dn.String()] = structs.TopologySourceRegistration
+	}
```

Comment:
awesome, this is roughly what I was thinking would be needed for trying to isolate some of this code